### PR TITLE
[Flash Attention 2] Add flash attention 2 for GPT-J

### PR DESCRIPTION
**Pull request by bytebarde**
# What does this PR do?

Adds Flash Attention 2 for `GPT-J` 
Fixes #[26350](https://github.com/huggingface/transformers/issues/26350)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc: @younesbelkada

### CODE REVIEW

**Comment by bytebarde** [comment]
Current progress with running `flash_attn_test`. Will dive deeper to fix the error.

**Comment by susnato** [comment]
Hi @bytebarde, what is the error message? 
If it is something like - "IndexError : tensors used as ...", then updating CUDA could solve the error (At least it was for my case).

**Comment by bytebarde** [comment]
Hi @susnato, thank you so much for for your attention to this PR!

I believe the error originates from two factors: (1) my preliminary implementation of `GPTJFlashAttention2`, which aimed to eliminate redundant transposing of the key and query, and (2) the execution of `test_flash_attn_2_generate_padding_right` using the testing configuration.

To address these issues, I have reinstated the original transposing operations and reverted the QKV cache concatenation. Additionally, I ran the test using the actual checkpoint, which successfully passed all related tests, as what  @younesbelkada and you did for llama2 and phi2.

**Comment by bytebarde** [comment]
Hi @younesbelkada,

I believe this pull request is now ready for your review.

I'd like to highlight a few changes, especially regarding `check_copies.py`, that I'm not entirely confident about. To ensure the branch passes the `make fixup` check, I removed the "copies" lines before both `modeling_codegen.CodeGenBlock` and `test_modeling_gptj.test_flash_attn_2_generate_padding_right`. This was done because the changes involved are somehow complex.

I would really appreciate your guidance on this. If there's a more standard or preferable way to handle such intricate changes, please let me know so I can make the necessary adjustments.

Thank you for your time on this!

File: tests/models/gptj/test_modeling_gptj.py [hunk 1]
```diff
@@ -518,6 +527,48 @@ def test_model_from_pretrained(self):
...
+    @slow
+    def test_flash_attn_2_generate_padding_right(self):
+        """
+        Overwritting the common test as the test is flaky on tiny models
+        """
+        model = GPTJForCausalLM.from_pretrained(
+            "EleutherAI/gpt-j-6b",
+            load_in_4bit=True,
+            device_map={"": 0},
+            revision="float16",
```
Comment (bytebarde) [inline tests/models/gptj/test_modeling_gptj.py:542]: The full version of GPT-J requires 48GB of RAM to load, so I used the float16 version provided by the model creators.

File: tests/models/gptj/test_modeling_gptj.py [hunk 2]
```diff
@@ -518,6 +527,48 @@ def test_model_from_pretrained(self):
...
+    @require_torch_gpu
+    @pytest.mark.flash_attn_test
+    @slow
+    def test_flash_attn_2_generate_padding_right(self):
+        """
+        Overwritting the common test as the test is flaky on tiny models
+        """
+        model = GPTJForCausalLM.from_pretrained(
+            "EleutherAI/gpt-j-6b",
+            load_in_4bit=True,
```
Comment (bytebarde) [inline tests/models/gptj/test_modeling_gptj.py:540]: The tests also passed without using `load_in_4bit=True`.

File: tests/models/gptj/test_modeling_gptj.py [hunk 3]
```diff
@@ -518,6 +527,48 @@ def test_model_from_pretrained(self):
             model = GPTJModel.from_pretrained(model_name, revision="float16", torch_dtype=torch.float16)
             self.assertIsNotNone(model)
 
+    @require_flash_attn
+    @require_torch_gpu
+    @pytest.mark.flash_attn_test
+    @slow
+    def test_flash_attn_2_generate_padding_right(self):
```
Comment (bytebarde) [inline tests/models/gptj/test_modeling_gptj.py:534]: removed copy lines here

File: src/transformers/models/codegen/modeling_codegen.py [hunk 4]
```diff
@@ -264,7 +264,6 @@ def forward(self, hidden_states: Optional[torch.FloatTensor]) -> torch.FloatTens
         return hidden_states
 
 
-# Copied from transformers.models.gptj.modeling_gptj.GPTJBlock with GPTJ->CodeGen
```
Comment (bytebarde) [inline src/transformers/models/codegen/modeling_codegen.py:267]: removed copy line here.

Comment (younesbelkada) [reply src/transformers/models/codegen/modeling_codegen.py:267]: I think it is better to keep it, you can either propagate the tiny changes the made in `GPTJAttention` there or put `# Ignore Copy` on top of the init method

File: tests/models/gptj/test_modeling_gptj.py (hunk 1 above)
Comment (younesbelkada) [reply tests/models/gptj/test_modeling_gptj.py:542]: Ok thanks!

**Review by younesbelkada — COMMENTED** [review]
Looks clean on my end already ! Would you be happy to address the comment about the copy mechanism that has been removed ? 
Also, can you run the benchmarking script here: https://gist.github.com/younesbelkada/02f35734da906cc0f2389ae4f665c58f with a gpt-j checkpoint and see the speedup (the result should give you something similar than: https://github.com/huggingface/transformers/pull/26414#pullrequestreview-1718072465), I can take care of pushing the images on the Hub and we'll just need to update the docs similarly as: https://github.com/huggingface/transformers/pull/26414/commits/7d4c688af58fdc43d423a7be6585e6c373937bdf

**Comment by bytebarde** [comment]
Hi @younesbelkada, thank you very much for your valuable input and guidance! I apologize for the delayed response.

I've addressed the comment regarding the copy mechanism, and the branch successfully passed the `make fixup` test.

Additionally, I've conducted the speed test. However, the observed speedup was not as significant as what we noted with OPT. The test was performed on an Nvidia RTX 4090, utilizing `max-batch-size=8` and `max-seqlen=32` to conserve memory. The model checkpoint used was `EleutherAI/gpt-j-6b` with the revision set to "float16". I've attached the speedup graph below for your review.






Could you also perform the test on an A100 GPU for comparison?

Thank you once again for your time. I look forward to hearing your thoughts on this!

**Review by younesbelkada — APPROVED** [review]
Thank you ! Can you just rebase / merge with main to make sure the CI passes?

File: src/transformers/models/gptj/modeling_gptj.py [hunk 5]
```diff
@@ -269,6 +292,250 @@ def forward(
...
+            value = torch.cat((past_value, value), dim=-2)
+
+        if use_cache is True:
+            # Note that this cast is quite ugly, but is not implemented before ROPE as the original codebase keeps the key in float32 all along the computation.
+            # Reference: https://github.com/kingoflolz/mesh-transformer-jax/blob/f8315e3003033b23f21d78361b288953064e0e76/mesh_transformer/layers.py#L128
+            present = (key.to(hidden_states.dtype), value)
+        else:
+            present = None
+
+        # The Falsh attention requires the input to have the shape
```
Comment (ArthurZucker) [inline src/transformers/models/gptj/modeling_gptj.py:378]:
```suggestion
        # The Flash attention requires the input to have the shape
```

File: src/transformers/models/gptj/modeling_gptj.py [hunk 6]
```diff
@@ -293,7 +560,11 @@ def __init__(self, config):
         super().__init__()
         inner_dim = config.n_inner if config.n_inner is not None else 4 * config.n_embd
         self.ln_1 = nn.LayerNorm(config.n_embd, eps=config.layer_norm_epsilon)
-        self.attn = GPTJAttention(config)
+        self.attn = (
```
Comment (ArthurZucker) [inline src/transformers/models/gptj/modeling_gptj.py:563]: let's define and use ! 
```suggestion
GPTJ_ATTENTION_CLASSES = {
    "eager": GPTJAttention,
    "flash_attention_2": GPTJFlashAttention,
}
```

**Review by ArthurZucker — APPROVED** [review]
LGTM Thanks for adding flash attention support!

**Review by younesbelkada — COMMENTED** [review]
Hi @bytebarde 
Can you address this comment? https://github.com/huggingface/transformers/pull/28295/files#r1470429885
It shouldn't be super hard, you just need to do something similar than what we do in Llama, specifically: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L746 and         
https://github.com/huggingface/transformers/blob/7b2bd1fbbd50e57cf28013e2d0737912ecc0f2eb/src/transformers/models/llama/modeling_llama.py#L758

**Comment by bytebarde** [comment]
Hi @ArthurZucker and @younesbelkada ,

Thank you so much for your additional suggestions!

I am sorry. I had assumed that `GPTJ_ATTENTION_CLASSES` had already been introduced by @ArthurZucker previously...

I have now added GPTJ_ATTENTION_CLASSES and made the necessary code modifications. 
Furthermore, I re-ran the test suite and successfully passed all the tests.

Please let me know if there's anything more I can do!
Thank you so much!

**Comment by ArthurZucker** [comment]
Good for me merging! 🤗

File: tests/models/gptj/test_modeling_gptj.py (hunk 3 above)
Comment (ArthurZucker) [inline tests/models/gptj/test_modeling_gptj.py:534]: requires_bitsandbytes here! 
Also let's add the expected text explicitly! to make sure we always have what we want!

**Review by ArthurZucker — APPROVED** [review]
Last nits!

**Comment by bytebarde** [comment]
Hi @ArthurZucker , good morning!

I have added the `@require_bitsandbytes` and expected outputs in the test function.
Please let me know if there is anything needed to be addressed.

Thank you so much!

**Review by younesbelkada — COMMENTED** [review]
Hi @bytebarde 
Thanks!  Can you run the styling checks? `make fixup` and / or `make fix-copies` after that we can merge

**Comment by bytebarde** [comment]
Hi @younesbelkada,

Thank you for taking the time to review this!

I have run the `make fix-copies` and believe that the previous issues regarding consistency have been addressed.

Please let me know if any further changes are needed. Thank you!

**Review by younesbelkada — APPROVED** [review]
Thanks again !

**Pull request merged**